File: src/test/kotlin/com/ericrochester/advent2022/TestDay11.kt
package com.ericrochester.advent2022

import kotlin.test.assertEquals
import kotlin.test.Ignore
import kotlin.test.Test
import kotlin.test.assertContains

class TestDay11 {
    private val inputData = javaClass.getResource("/examples/day11.txt")?.readText() ?: ""
    private val day = Day11()

    @Test fun testDay11RunA() {
        val output = day.runA(inputData)
        assertEquals(10605, output)
    }

    @Ignore @Test fun testDay11RunB() {
        val output = day.runB(inputData)
        assertEquals(TODO(), output)
    }
}

class TestMonkey {
    val monkey = Monkey(
        1,
        mutableListOf(54, 65, 75, 74),
        OpPair.plus(6),
        OpPair.divisible(19),
        ThrowTo(2, 0)
    )

    @Test fun testParse() {
        val input = """
            |Monkey 1:
            |  Starting items: 54, 65, 75, 74
            |  Operation: new = old + 6
            |  Test: divisible by 19
            |    If true: throw to monkey 2
            |    If false: throw to monkey 0
            |
        """.trimMargin()
        val expected = monkey

        assertEquals(expected, Monkey.parse(input.lines()))
    }

    @Test fun testParseOldOld() {
        val input = """
            |Monkey 2:
            |  Starting items: 79, 60, 97
            |  Operation: new = old * old
            |  Test: divisible by 13
            |    If true: throw to monkey 1
            |    If false: throw to monkey 3
        """.trimMargin()
        val expected = Monkey(
            2,
            mutableListOf(79, 60, 97),
            OpPair.pow(2),
            OpPair.divisible(13),
            ThrowTo(1, 3)
        )
        assertEquals(expected, Monkey.parse(input.lines()))
    }

    @Test fun testMove() {
        val expected = listOf(
            ThrowInfo(0, 20),
            ThrowInfo(0, 23),
            ThrowInfo(0, 27),
            ThrowInfo(0, 26)
        )
        assertEquals(expected, monkey.move())
        assertEquals(4, monkey.itemsSeen)
    }

    @Test fun testTracksItemsSeen() {
        monkey.catchItem(42)
        assertContains(monkey.items, 42)
    }
}

File: advent2022/test_day11.py

from advent2022 import day11


def test_run_a():
    input_file = './src/test/resources/examples/day11.txt'
    with open(input_file) as f:
        output = day11.run_a(f)
    assert(10605 == output)


def test_run_b():
    input_file = './src/test/resources/examples/day11.txt'
    with open(input_file) as f:
        output = day11.run_b(f)
    assert(2713310158 == output)


MONKEY1 = '''\
Monkey 1:
  Starting items: 54, 65, 75, 74
  Operation: new = old + 6
  Test: divisible by 19
    If true: throw to monkey 2
    If false: throw to monkey 0
'''


def test_parse_monkey():
    monkey = day11.Monkey.parse(MONKEY1.splitlines(), 3)
    assert(monkey.n == 1)
    assert(list(monkey.items) == [54, 65, 75, 74])
    assert(monkey.op == (day11.Op.PLUS, 6))
    assert(monkey.test == (day11.Op.DIVISIBLE_BY, 19))
    assert(monkey.action == (2, 0))


def test_monkey_process():
    monkey = day11.Monkey.parse(MONKEY1.splitlines(), 3)
    output = monkey.inspect(54)
    assert(output == (0, 20))
    assert(monkey.inspected == 1)


def test_op_parse():
    tests = [
        ('new = old + 6', (day11.Op.PLUS, 6)),
        ('divisible by 19', (day11.Op.DIVISIBLE_BY, 19)),
        ('new = old * 19', (day11.Op.MULTIPLY, 19)),
        ('new = old * old', (day11.Op.POWER, 2)),
    ]
    for (input_str, expected) in tests:
        assert(day11.Op.parse(input_str) == expected)


def test_op_exec():
    tests = [
        (day11.Op.PLUS, 3, 2, 5),
        (day11.Op.MULTIPLY, 7, 4, 28),
        (day11.Op.SUBTRACT, 7, 4, 3),
        (day11.Op.DIVIDE, 8, 4, 2),
        (day11.Op.DIVISIBLE_BY, 8, 4, 1),
        (day11.Op.DIVISIBLE_BY, 8, 3, 0),
        (day11.Op.POWER, 7, 2, 49),
    ]
    for (op, worry, value, expected) in tests:
        assert(day11.Op.exec(op, worry, value) == expected)